src/modules/flow/netctl/netctl.c

```diff
@@ -0,0 +1,336 @@
+        if (name && strcmp(name, service_name) == 0)
+            return service;
+    }
+
+    return NULL;
+}
+
+static int
+connection_handle(struct sol_flow_node *node, void *data, bool connect)
+{
+    struct network_service_data *mdata = data;
+    struct sol_netctl_service *service;
+    int r;
+
+    service = find_service_by_name(mdata->service_name);
+    if (!service) {
+        sol_flow_send_error_packet(node, ENOENT, "Did not found service name %s", mdata->service_name);
+        return 0;
+    }
+
+    if (connect) {
+       r = sol_netctl_service_connect(service);
+       if (r < 0)
+            sol_flow_send_error_packet(node, -r, "Could not connect to service: %s: %s",
+                     mdata->service_name, sol_util_strerrora(-r));
+    } else {
+       r = sol_netctl_service_disconnect(service);
+       if (r < 0)
+            sol_flow_send_error_packet(node, -r, "Could not disconnect to service: %s: %s",
+                     mdata->service_name, sol_util_strerrora(-r));
+    }
+
+    return 0;
+}
+
+static int
+connect_service_name(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    return connection_handle(node, data, true);
+}
+
+static int
+disconnect_service_name(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    return connection_handle(node, data, false);
+}
+
+static int
+get_service_state(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    struct network_service_data *mdata = data;
+    struct sol_netctl_service *service;
+    int r;
+
+    service = find_service_by_name(mdata->service_name);
+    if (!service) {
+        sol_flow_send_error_packet(node, ENOENT, "Did not found service name %s", mdata->service_name);
+        return 0;
+    }
+    r = sol_flow_send_string_packet(node,
+             SOL_FLOW_NODE_TYPE_NETCTL_SERVICE__OUT__OUT,
+             sol_netctl_service_state_to_str(mdata->state));
+    SOL_INT_CHECK(r, < 0, r);
+
+    if (mdata->state == SOL_NETCTL_SERVICE_STATE_ONLINE)
+            r = sol_flow_send_bool_packet(node,
+                         SOL_FLOW_NODE_TYPE_NETCTL_SERVICE__OUT__ONLINE,
+                         true);
+    else
+            r = sol_flow_send_bool_packet(node,
+                         SOL_FLOW_NODE_TYPE_NETCTL_SERVICE__OUT__ONLINE,
+                         false);
+
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+static int
+set_service_name(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    struct network_service_data *mdata = data;
+    const char *in_value;
+    int r;
+
+    r = sol_flow_packet_get_string(packet, &in_value);
+    SOL_INT_CHECK(r, < 0, r);
+
+    mdata->service_name = strdup(in_value);
+    SOL_NULL_CHECK(mdata->service_name, -ENOMEM);
+
```

Comment:
Consider the following conditions and the given FBP:

```sh
# when user changes the service name, update the one we monitor
some_user_input OUT -> NAME service
# show the monitor to user
service STATE -> IN some_user_output
```

## Service did not exist, then appeared

In this scenario, when `some_user_input OUT` generates a string packet `"someAP"`, the connman list did **not** know about any "someAP".

The `"someAP"` packet is processed by `set_service_name` and stored.

**After** the `"someAP"` packet is processed, connman finds a new service (ie: manual Scan DBus method, periodic scan...), and sends a DBus Signal "ServicesChanged", it will call our `service_status_cb()`, which will see `strcmp(mdata->service_name, name)` that matches and will send a new status on `service STATE -> IN some_user_output`. This will print out the output to user.

## Service did exist before

In this scenario, when `some_user_input OUT` generates a string packet `"someAP"`, then connman and `sol-netctl` did know about "someAP".

The `"someAP"` packet is processed by `set_service_name` and stored.

In your code, that stops there. No packets will be ever emitted in `service STATE -> IN some_user_output`, since that is only done by a change in `service_status_cb()` or if the user manually send some packet to `TRIGGER service`.

With my proposal, inside `set_service_name()`, before you return you should call `get_service_state()`. In this case it would call `sol_netctl_get_services_list()` and note that `mdata->service_name` (new) is there, thus will emit the `service STATE -> IN some_user_output` that we need.